File: src/main/java/io/stouder/adonis/service/AdonisAceService.java
package io.stouder.adonis.service;

import com.intellij.openapi.project.Project;
import io.stouder.adonis.cli.json.ace.Command;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Consumer;

public interface AdonisAceService {
    static AdonisAceService getInstance(Project project) {
        return project.getService(AdonisAceService.class);
    }

    <T> Map<String, Optional<T>> runAceGetCommandOnEveryRoots(String progressTitle, List<String> parameters, Class<T> responseType);
    <T> void runAceGetCommandAsyncOnEveryRoots(Consumer<Map<String, Optional<T>>> callback, List<String> parameters, Class<T> responseType);
    boolean execAceCommand(String progressTitle, List<String> parameters, String basePath);
    void fetchCommands(Consumer<Map<String, Optional<Command[]>>> callback);
}

File: src/main/java/io/stouder/adonis/service/AdonisAceService.kt
package io.stouder.adonis.service

import com.intellij.openapi.project.Project
import io.stouder.adonis.cli.json.ace.Command
import java.util.*
import java.util.function.Consumer

interface AdonisAceService {
    companion object {
        fun getInstance(project: Project): AdonisAceService {
            return project.getService(AdonisAceService::class.java)
        }
    }

    fun <T> runAceGetCommandOnEveryRoots(progressTitle: String, parameters: List<String>, responseType: Class<T>): Map<String, Optional<T & Any>>
    fun <T> runAceGetCommandAsyncOnEveryRoots(callback: Consumer<Map<String, Optional<T & Any>>>, parameters: List<String>, responseType: Class<T>)
    fun execAceCommand(progressTitle: String, parameters: List<String>, basePath: String): Boolean
    fun fetchCommands(callback: Consumer<Map<String, Optional<Array<Command>>>>)
}
